### PR TITLE
HPCC-22224 Document Dali Whitelist mechanism

### PR DESCRIPTION
**Pull request by JamesDeFabia**
Signed-off-by: Jim DeFabia <jamesdefabia@lexisnexis.com>
@g-pan Please review for grammar, style, and clarity
@jakesmith Please review  (Link to pdf below)
@kenrowland Please review

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

PDF: 
http://10.240.32.243/view/Docs%207.0/job/JimDocBuild7/CE=ubuntu-18.04-amd64-1/lastSuccessfulBuild/artifact/build-docs/Release/docs/EN_US/HPCCSystemAdministratorsGuide_EN_US-7.3.0-trunk0.pdf#page=117

### CODE REVIEW

File: docs/EN_US/HPCCSystemAdmin/SA-Mods/DaliLDAP.xml [hunk 1]
```diff
@@ -97,4 +97,174 @@
       with File Scope Security enabled.</para>
     </sect3>
   </sect2>
+
+  <sect2 id="WhitelistInDali" role="brk">
+    <title>Whitelist in Dali</title>
+
+    <para>The Dali server has the ability to restrict access to only those
+    nodes that are associated with a role in the environment definition
+    (environment.xml) or explicitly added to the Whitelist in Dali's
```
Comment (g-pan) [inline docs/EN_US/HPCCSystemAdmin/SA-Mods/DaliLDAP.xml]: add emphasis to first instance of Whitelist

File: docs/EN_US/HPCCSystemAdmin/SA-Mods/DaliLDAP.xml [hunk 2]
```diff
@@ -97,4 +97,174 @@
...
+    <para>Nodes can have one or more roles, specified in a comma separated
+    list.</para>
+
+    <para>To disable the Whitelist feature entirely, you can add :</para>
+
+    <para><programlisting>&lt;WhiteList enabled="false"&gt;</programlisting><emphasis
+    role="bold">This is not recommended for production
+    environments.</emphasis></para>
+
+    <para><?hard-pagebreak ?></para>
```
Comment (g-pan) [inline docs/EN_US/HPCCSystemAdmin/SA-Mods/DaliLDAP.xml]: remove line 196

File: docs/EN_US/HPCCSystemAdmin/SA-Mods/DaliLDAP.xml [hunk 3]
```diff
@@ -97,4 +97,174 @@
...
+
+    <sect3>
```
Comment (g-pan) [inline docs/EN_US/HPCCSystemAdmin/SA-Mods/DaliLDAP.xml]: add role="brk" to sect3

File: docs/EN_US/HPCCSystemAdmin/SA-Mods/DaliLDAP.xml [hunk 4]
```diff
@@ -97,4 +97,174 @@
...
+      <title role="brk">Use envmod to add to the supplementary
```
Comment (g-pan) [inline docs/EN_US/HPCCSystemAdmin/SA-Mods/DaliLDAP.xml]: remove role="brk" from title

File: docs/EN_US/HPCCSystemAdmin/SA-Mods/DaliLDAP.xml [hunk 5]
```diff
@@ -97,4 +97,174 @@
...
+      Whitelist</title>
+
+      <para>We recommend using the <emphasis role="bold">envmod</emphasis>
+      command line tool to add Whitelist entries to your environment.xml file.
+      The envmod utility can be found in /opt/HPCCSystems/bin/.</para>
+
+      <para>Use a template file to such as the following example:</para>
```
Comment (g-pan) [inline docs/EN_US/HPCCSystemAdmin/SA-Mods/DaliLDAP.xml]: remove "to"

**Review by g-pan — COMMENTED** [review]
see comments in line

**Comment by jakesmith** [comment]
@JamesDeFabia - looks good. The only question mark is that there is no explanation of why you'd want to add additional explicit WhiteList roles in the environment.

Listing all the role types gives the impression that it might be legitimate to add new ThorMaster's or SashaServer's for example, when in reality that should never be necessary because they must already be part of the environment and therefore implicitly added.

The only roles that are likely to be added are tools and utilities that are not defined components, and are run by ops/admin (probably from outside of the environment). i.e. Config, DaFsControl, DaliAdmin, DaliDiag, SchedulerAdmin, SwapNode, Testing, TreeView, UpdateEnv, XRef

Perhaps a section that says something like this:

NB: It should never be necessary to explicitly add the following roles : 
```AgentExec, DaliServer, EclAgent, EclCC, EclCCServer, EclServer, EspServer, SashaServer, RoxieMaster, ThorMaster```

These roles correspond to components in the environment, which should always be implicitly white listed, therefore any additional components of these types should be added to the environment, and therefore also implicitly added to the white list.

Also note by default the following roles are automatically white listed to run on the Dali node:
```Config, DaFsControl, DaliAdmin, DaliDiag, ScheduleAdmin, SwapNode, Testing, TreeView, UpdateEnv, XRef```

i.e. by default they can only be run from the Dali node.

**Comment by jakesmith** [comment]
@JamesDeFabia - I've just added a new role "Monitoring" that was needed by our Nagios monitoring. It is in the same category as other admin/ops. roles above.
PR that adds: https://track.hpccsystems.com/browse/HPCC-22356

**Comment by JamesDeFabia** [comment]
Updated, rebased, and rebuilt the PDF
@g-pan Please review
@jakesmith Please have another look
@kenrowland Please review now

**Review by g-pan — COMMENTED** [review]
See comments inline

**Comment by jakesmith** [comment]
@JamesDeFabia - looks good.

**Comment by kenrowland** [comment]
@JamesDeFabia  looks accurate

**Comment by JamesDeFabia** [comment]
@g-pan I made the minor edits you requested, please give this a final look

**Review by g-pan — APPROVED** [review]
looks good now.

**Comment by JamesDeFabia** [comment]
@richardkchapman This is good to merge. Greg had the last review for some minor edits, but @jakesmith  and @kenrowland had already approved technical content

**Pull request merged**